robowflex_library/include/robowflex_library/io.h

```diff
@@ -39,43 +39,51 @@ namespace robowflex
          *  \param[in] path Path to resolve.
          *  \return The canonical path, or "" on failure.
          */
-        const std::string resolvePath(const std::string &path);
+        std::string resolvePath(const std::string &path);
 
         /** \brief Resolves `package://` URLs to get the directory this path is in.
          *  \param[in] path Path to get the parent of.
          *  \return The directory that this path is contained in, or "" on failure.
          */
-        const std::string resolveParent(const std::string &path);
+        std::string resolveParent(const std::string &path);
+
+        /** \brief Concatenates two elements of a path, a directory and a filename.
+         *  \param[in] directory Path to use as directory. If there are other elements at the end of the
+         *                       directory path, they will be removed.
+         *  \param[in] filename Filename to add.
+         *  \return The canonical path for this path.
+         */
+        std::string makeFilepath(const std::string &directory, const std::string &filename);
 
         /** \brief Loads an XML or .xacro file to a string.
          *  \param[in] path File to load.
          *  \return The loaded file, or "" on failure (file does not exist or .xacro is malformed).
          */
-        const std::string loadXMLToString(const std::string &path);
+        std::string loadXMLToString(const std::string &path);
 
         /** \brief Loads a .xacro file to a string.
          *  \param[in] path File to load.
          *  \return The loaded file, or "" on failure (file does not exist or .xacro is malformed).
          */
-        const std::string loadXacroToString(const std::string &path);
+        std::string loadXacroToString(const std::string &path);
 
         /** \brief Loads a file to a string.
          *  \param[in] path File to load.
          *  \return The loaded file, or "" on failure (file does not exist).
          */
-        const std::string loadFileToString(const std::string &path);
+        std::string loadFileToString(const std::string &path);
 
         /** \brief Runs a command \a cmd and returns stdout as a string.
          *  \param[in] cmd Command to run.
          *  \return Contents of stdout from \a cmd, or "" on failure.
          */
-        const std::string runCommand(const std::string &cmd);
+        std::string runCommand(const std::string &cmd);
 
         /** \brief Loads a file to a YAML node.
          *  \param[in] path File to load.
          *  \return A pair, where the first is true on success false on failure, and second is the YAML node.
          */
-        const std::pair<bool, YAML::Node> loadFileToYAML(const std::string &path);
+        std::pair<bool, YAML::Node> loadFileToYAML(const std::string &path);
```

Comment:
Maybe this should stay const since it could load big files.

Reply:
The const doesn't matter here since it only applies to the pair and we can get rid of it. If we want to avoid copies no matter what the API would have to be changed, but if the caller uses something like `const auto &pair = loadFileToYAML(...)` the compiler should be smart enough to not do the copy.